### PR TITLE
Fix stringop-truncation compiling with GCC 9.3

### LINKED ISSUE #31717: Fix stringop-truncation compiling with GCC 9.3
```
/home/sam/mantid/Framework/Kernel/src/NexusHDF5Descriptor.cpp:109:12: warning: 'char* strncpy(char*, const char*, size_t)' specified bound 128 equals destination size [-Wstringop-truncation]
  109 |     strncpy(data, vdat, maxlen);
      |     ~~~~~~~^~~~~~~~~~~~~~~~~~~~
```
When compiling with GCC 9.3 notably in a Conda environment, this is an issue, when compiling and needs to be solved before a switch happens.

### PR DESCRIPTION
**Pull request by Pasarus**
**Description of work.**
There is a warning when compiling with GCC 9.3, this fixes that warning.

**To test:**
CI passes and code review. 

Fixes #31717 

*This does not require release notes* because **Developer Facing Only**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Kernel/src/NexusHDF5Descriptor.cpp [hunk 1]
```diff
@@ -100,16 +100,16 @@ herr_t readStringAttribute(hid_t attr, char **data) {
  * @param maxlen
```
Comment (joseph-torsney) [inline Framework/Kernel/src/NexusHDF5Descriptor.cpp:100]:
```suggestion
```

**Review by joseph-torsney — CHANGES_REQUESTED** [review]
Code appears to behave the same, and CI passes. Just one tiny change to code comment :)

**Review by joseph-torsney — DISMISSED** [review]

File: Framework/Kernel/src/NexusHDF5Descriptor.cpp [hunk 2]
```diff
@@ -97,19 +97,18 @@ herr_t readStringAttribute(hid_t attr, char **data) {
  * Reads a string attribute of N-dimensions
  * @param attr input HDF5 attribute handler
  * @param data output attribute data
- * @param maxlen
  * @return
  */
-herr_t readStringAttributeN(hid_t attr, char *data, int maxlen) {
+std::pair<std::string, herr_t> readStringAttributeN(hid_t attr, char *data) {
```
Comment (martyngigg) [inline Framework/Kernel/src/NexusHDF5Descriptor.cpp:102]: Do we need `data` ta all now? `dataString` is assigned to completely anyway? Can we be more direct and :
```c++
std::pair<std::string, herr_t> readStringAttributeN(hid_t attr) {
  char *vdat = NULL;
  const auto iRet = readStringAttribute(attr, vdat);
  std::string attrData;
  if(iRet >= 0) {
    attrData = vDat;
  }
  return std::make_pair(attrData, iRet);
}
```

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: Framework/Kernel/src/NexusHDF5Descriptor.cpp (hunk 2 above)
Comment (Pasarus) [reply Framework/Kernel/src/NexusHDF5Descriptor.cpp:102]: If `iRet is not >=0` then `dataString != vdat` and the original data value is returned which was the previous behavior, in your case we don't return the original input anymore if `iRet is not >=0`. I tried to be as unobtrusive and preserve the original behavior here as I don't want to change some very old behavior in our code base, just to fix this warning.

**Review by martyngigg — APPROVED** [review]
Thanks. I think this is a bit cleaner now.

**Pull request merged**